xarray/core/indexing.py

```diff
@@ -277,84 +278,224 @@ def _index_indexer_1d(old_indexer, applied_indexer, size):
+                if k.ndim != 1:
+                    raise TypeError('invalid indexer array for {}, must have '
+                                    'exactly 1 dimension: '
+                                    .format(type(self).__name__, k))
+                k = np.asarray(k, dtype=np.int64)
+            else:
+                raise TypeError('unexpected indexer type for {}: {!r}'
+                                .format(type(self).__name__, k))
+            new_key.append(k)
+
+        super(OuterIndexer, self).__init__(new_key)
+
+
+class VectorizedIndexer(ExplicitIndexer):
+    """Tuple for vectorized indexing.
+
+    All elements should be slice or N-dimensional np.ndarray objects with an
+    integer dtype. Indexing follows proposed rules for np.ndarray.vindex, which
+    matches NumPy's advanced indexing rules (including broadcasting) except
+    sliced axes are always moved to the end:
+    https://github.com/numpy/numpy/pull/6256
+    """
+    def __init__(self, key):
+        if not isinstance(key, tuple):
+            raise TypeError('key must be a tuple: {!r}'.format(key))
+
+        new_key = []
+        for k in key:
+            if isinstance(k, slice):
+                k = as_integer_slice(k)
+            elif isinstance(k, np.ndarray):
+                if not np.issubdtype(k.dtype, np.integer):
+                    raise TypeError('invalid indexer array, does not have '
+                                    'integer dtype: {!r}'.format(k))
+                k = np.asarray(k, dtype=np.int64)
+            else:
+                raise TypeError('unexpected indexer type for {}: {!r}'
+                                .format(type(self).__name__, k))
+            new_key.append(k)
+
+        super(VectorizedIndexer, self).__init__(new_key)
+
+
+class ExplicitlyIndexed(object):
+    """Mixin to mark support for Indexer subclasses in indexing."""
 
 
-class VectorizedIndexer(IndexerTuple):
-    """ Tuple for vectorized indexing """
+class ExplicitNDArrayMixin(utils.NDArrayMixin, ExplicitlyIndexed):
 
+    def __array__(self, dtype=None):
+        key = BasicIndexer((slice(None),) * self.ndim)
+        return np.asarray(self[key], dtype=dtype)
+
+
+def unwrap_explicit_indexer(key, target, allow):
+    """Unwrap an explicit key into a tuple."""
+    if not isinstance(key, ExplicitIndexer):
+        raise TypeError('unexpected key type: {}'.format(key))
+    if not isinstance(key, allow):
+        key_type_name = {
+            BasicIndexer: 'Basic',
+            OuterIndexer: 'Outer',
+            VectorizedIndexer: 'Vectorized'
+        }[type(key)]
+        raise NotImplementedError(
+            '{} indexing for {} is not implemented. Load your data first with '
+            '.load(), .compute() or .persist(), or disable caching by setting '
+            'cache=False in open_dataset.'.format(key_type_name, type(target)))
+    return key.tuple
+
+
+class WrapImplicitIndexing(utils.NDArrayMixin):
```

Comment:
This name sounds like a function which behaves opposite to `unwrap_explicit_indexer`. Can we change this something like `ImplicitIndexedArrayWrapper`?
(But as the below comment, I did not yet understand how this wrapper will be used.)